src/olympia/blocklist/tests/test_mlbf.py

```diff
@@ -300,19 +311,217 @@ def test_diff_block_added_and_removed(self):
+                set(
+                    MLBF.hash_filter_inputs(
+                        [(block_version.block.guid, block_version.version.version)]
+                    )
+                ),
+                2,
             ),
-            set(
-                MLBF.hash_filter_inputs(
-                    [(block_version.block.guid, block_version.version.version)]
-                )
+            BlockType.SOFT_BLOCKED: (set(), set(), 0),
+        }
+
+    def test_diff_block_hard_to_soft(self):
+        addon, block = self._blocked_addon()
+        block_version = self._block_version(
+            block, self._version(addon), block_type=BlockType.BLOCKED
+        )
+        base_mlbf = MLBF.generate_from_db('test')
+        block_version.update(block_type=BlockType.SOFT_BLOCKED)
+        next_mlbf = MLBF.generate_from_db('test_two')
+
+        assert next_mlbf.generate_diffs(previous_mlbf=base_mlbf) == {
+            BlockType.BLOCKED: (
+                set(),
+                set(
+                    MLBF.hash_filter_inputs(
+                        [(block_version.block.guid, block_version.version.version)]
+                    )
+                ),
+                1,
+            ),
+            BlockType.SOFT_BLOCKED: (
+                set(
+                    MLBF.hash_filter_inputs(
+                        [(block_version.block.guid, block_version.version.version)]
+                    )
+                ),
+                set(),
+                1,
+            ),
+        }
+
+    def test_diff_block_soft_to_hard(self):
+        addon, block = self._blocked_addon()
+        block_version = self._block_version(
+            block, self._version(addon), block_type=BlockType.SOFT_BLOCKED
+        )
+        base_mlbf = MLBF.generate_from_db('test')
+        block_version.update(block_type=BlockType.BLOCKED)
+        next_mlbf = MLBF.generate_from_db('test_two')
+
+        assert next_mlbf.generate_diffs(previous_mlbf=base_mlbf) == {
+            BlockType.BLOCKED: (
+                set(
+                    MLBF.hash_filter_inputs(
+                        [(block_version.block.guid, block_version.version.version)]
+                    )
+                ),
+                set(),
+                1,
+            ),
+            BlockType.SOFT_BLOCKED: (
+                set(),
+                set(
+                    MLBF.hash_filter_inputs(
+                        [(block_version.block.guid, block_version.version.version)]
+                    )
+                ),
+                1,
             ),
-            2,
+        }
+
+    def test_diff_all_possible_changes(self):
+        """
+        Simulates 6 guid:version comibinations moving from one state to another
+        covering all possible movements and the final diff state.
+
+        1. Not blocked -> Soft blocked
+        2. Not blocked -> Blocked
+        3. Soft blocked -> Blocked
+        4. Soft blocked -> Not blocked
+        5. Blocked -> Soft blocked
```

Comment:
Use of the word this should be forbidden in code reviews. Which this are you referring to here?

Each one of these edge cases is important to cover for either modern or legacy versions of FX and moreover we have to ensure the backward compatibility that we agreed on (that you proposed) does not break either legacy or modern Fx clients.

Reply:
_This_ line 😬  Not sure if that makes the code more complex or anything, but there won't be hard to soft block transitions.

Reply:
There is a PR introducing that feature (hard->soft) now so AFAIK yes it will be possible to do. Even if we don't use hard->soft feature I still think it makes sense to account for the total possible transitions, MLBF shouldn't care about how we intend to use the feature it just works.. esp since the code covering this case is already here and tested.

hard to soft results in a version being in both unblocked and soft-blocked on the next stash

Reply:
Ok, that's fine.